src/Neo.VM/Types/Array.cs

```diff
@@ -35,7 +35,15 @@ public class Array : CompoundType, IReadOnlyList<StackItem>
                 if (IsReadOnly) throw new InvalidOperationException("The object is readonly.");
                 ReferenceCounter?.RemoveReference(_array[index], this);
                 _array[index] = value;
-                ReferenceCounter?.AddReference(value, this);
+                if (value is CompoundType { ReferenceCounter: null } compoundType)
+                {
+                    compoundType.ReferenceCounter = ReferenceCounter;
+                    ReferenceCounter?.AddReference(value.DeepCopy(), this);
```

Comment:
I fear this can explode, imagine deeply nested structures that you would try to deep-copy just to count references, what's gonna happen to execution time with this code?

Reply:
if deepcopy can go explode, then i think it should be a problem of deepcopy, isnt it?

Reply:
>  i think it should be a problem of deepcopy

It will be a problem of the whole node.

Reply:
Think of things like #2723. DeepCopy will do its job, no doubt about it, but it will affect the cost/time balance of some opcodes. Like PICKITEM is affected by this line IIUC and it's cheap, it's supposed to be cheap.

Reply:
> > i think it should be a problem of deepcopy
> 
> It will be a problem of the whole node.

I mean if deepcopy can explode, it means deepcopy itself is problematic, and should be fixing deepcopy, instead of here. And @roman-khimov's concern is reasonable, but if balance is a concern here, it should be a concern everywhere and everytime we make any change to the core. We dont have benchmark for them, has being working like this. But I will add a benchmark specifically for this one.